pkg/ottl/ottlfuncs/func_slice_to_map.go

```diff
@@ -32,80 +32,107 @@ func sliceToMapFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ot
 	return getSliceToMapFunc(args.Target, args.KeyPath, args.ValuePath), nil
 }
 
-func getSliceToMapFunc[K any](target ottl.Getter[K], keyPath, valuePath ottl.Optional[[]string]) ottl.ExprFunc[K] {
+func getSliceToMapFunc[K any](target ottl.PSliceGetter[K], keyPath, valuePath ottl.Optional[[]string]) ottl.ExprFunc[K] {
 	return func(ctx context.Context, tCtx K) (any, error) {
 		val, err := target.Get(ctx, tCtx)
 		if err != nil {
 			return nil, err
 		}
 
-		switch v := val.(type) {
-		case []any:
-			return sliceToMap(v, keyPath, valuePath)
-		case pcommon.Slice:
-			return sliceToMap(v.AsRaw(), keyPath, valuePath)
-		default:
-			return nil, fmt.Errorf("unsupported type provided to SliceToMap function: %T", v)
-		}
+		return sliceToMapFromPcommon(val, keyPath, valuePath)
 	}
 }
 
-func sliceToMap(v []any, keyPath, valuePath ottl.Optional[[]string]) (any, error) {
+func sliceToMapFromPcommon(v pcommon.Slice, keyPath, valuePath ottl.Optional[[]string]) (pcommon.Map, error) {
 	m := pcommon.NewMap()
-	m.EnsureCapacity(len(v))
-	for i, elem := range v {
-		e, ok := elem.(map[string]any)
-		if !ok {
-			return nil, fmt.Errorf("could not cast element '%v' to map[string]any", elem)
+	m.EnsureCapacity(v.Len())
+
+	useKeyPath := !keyPath.IsEmpty()
+	useValuePath := !valuePath.IsEmpty()
+
+	for i := 0; i < v.Len(); i++ {
+		elem := v.At(i)
+
+		// If key_path is not set, key is the index
+		key := strconv.Itoa(i)
+		// If value_path is not set, value is the whole element
+		value := elem
+
+		// If using key_path or value_path, element must be a map
+		if useKeyPath || useValuePath {
+			if elem.Type() != pcommon.ValueTypeMap {
+				return pcommon.Map{}, fmt.Errorf("slice elements must be maps when using `key_path` or 'value_path', but could not cast element '%s' to a map", elem.Str())
```

Comment:
```suggestion
				return pcommon.Map{}, fmt.Errorf("slice elements must be maps when using \"key_path\" or \"value_path\", but could not cast element of type %q to a map", elem.Type())
```

`elem.Str()` returns the string value associated with this `pccomon.Value`, it might not be a string.